docs/building_the_project_with_dpcpp.rst

```diff
@@ -0,0 +1,470 @@
+external libraries. See the section :ref:`building_and_running_tests` for more
+information.
+
+The most important supported build options are:
+
+.. list-table::
+   :header-rows: 1
+
+   * - CMake Option
+     - Supported Values
+     - Default Value 
+   * - ENABLE_MKLCPU_BACKEND
+     - True, False
+     - True      
+   * - ENABLE_MKLGPU_BACKEND
+     - True, False
+     - True      
+   * - ENABLE_CUBLAS_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_CUSOLVER_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_CUFFT_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_CURAND_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_NETLIB_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_ROCBLAS_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_ROCFFT_BACKEND
+     - True, False
+     - False    
+   * - ENABLE_ROCSOLVER_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_ROCRAND_BACKEND
+     - True, False
+     - False     
+   * - ENABLE_MKLCPU_THREAD_TBB
+     - True, False
+     - True      
+   * - ENABLE_PORTBLAS_BACKEND
+     - True, False
+     - False      
+   * - ENABLE_PORTFFT_BACKEND
+     - True, False
+     - False      
+   * - BUILD_FUNCTIONAL_TESTS
+     - True, False
+     - True      
+   * - BUILD_EXAMPLES
+     - True, False
+     - True      
+   * - TARGET_DOMAINS (list)
+     - blas, lapack, rng, dft, sparse_blas
+     - All domains 
+
+Some additional build options are given in the section `Additional build options`_.
+
+.. _build_target_domains:
+
+TARGET_DOMAINS
+^^^^^^^^^^^^^^
+
+OneMKL supports multiple domains: BLAS, DFT, LAPACK, RNG and sparse BLAS. The
+domains built by oneMKL can be selected using the ``TARGET_DOMAINS`` parameter.
+In most cases, ``TARGET_DOMAINS`` is set automatically according to the domains
+supported by the backend libraries enabled. However, while most backend
+libraries support only one of these domains, but some may support multiple. For
+example, the ``MKLCPU`` backend supports every domain. To enable support for
+only the BLAS domain in the oneMKL Interfaces whilst compiling with ``MKLCPU``,
+``TARGET_DOMAINS`` could be set to ``blas``. To enable BLAS and DFT,
+``-DTARGET_DOMAINS="blas dft"`` would be used.
+
+
+Backends
+#########
+
+.. _build_for_intel_onemkl_dpcpp:
+
+Building for Intel(R) oneMKL
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Intel(R) oneMKL backend supports multiple domains on both x86 CPUs and Intel
+GPUs. The MKLCPU backend using Intel(R) oneMKL for x86 CPU is enabled by
+default, and controlled with the parameter ``ENABLE_MKLCPU_BACKEND``. The MKLGPU
+backend using Intel(R) oneMKL for Intel GPU is enabled by default, and
+controlled with the parameter ``ENABLE_MKLGPU_BACKEND``.
+
+When using the Intel(R) oneAPI DPC++ Compiler, it is likely that Intel(R) oneMKL
+will be found automatically. If it is not, the parameter ``MKL_ROOT`` can be set
```

Comment:
I am not so sure about this. It could depend on the installed package, like toolkit, standalone, or even pip, although I don't know if it will work.

Reply:
[The CMake here](https://github.com/oneapi-src/oneMKL/blob/7566579b1723e0bf932d9c9af8448acdc5267ea3/CMakeLists.txt#L251) suggests to me that `MKL_ROOT` can always be given by the user as a hint for finding MKL.

Is your preference that I suggest the usage of setting the environement variable `MKLROOT` instead, or that I suggest both?

Reply:
I think let's suggest `MKL_ROOT`, which would be the least possible to have surprises, then mention it can be omitted if `MKLROOT` is set either manually or through an environment script provided by the package.

Reply:
Done in [c55db76](https://github.com/oneapi-src/oneMKL/pull/510/commits/c55db7633e5a60d2abc56826deeb1eadc394f151).

Reply:
Sorry I didn't make it clear that it is "When using the Intel(R) oneAPI DPC++ Compiler, it is likely that Intel(R) oneMKL
will be found automatically." I am not sure about.

We can fine-tune it later. [The latest build](https://github.com/oneapi-src/oneMKL/actions/runs/9578987348) looks good to me. Let's merge this version.

Reply:
I guess it depends on how the base toolkit is being used? I use the `setvars.sh` script, after which MKL is found automagically. Likewise, on machines where `module load mkl` is needed, no manual intervention seems to be needed. I've used `MKL_ROOT` before, but I can't remember why.

Thanks! It might also be that we get feedback from users that can also help us fine tune things.

Reply:
>It might also be that we get feedback from users that can also help us fine tune things.

Good idea!

>I guess it depends on how the base toolkit is being used?

Indeed. For the base toolkit, which includes the compiler and MKL, the `setvars.sh` script will call each component's `env/vars.sh`, which sets `CMPLR_ROOT`, `MKLROOT`, etc. Hence MKL is found because of `MKLROOT`. If users do not install the base toolkit but install the compiler and MKL individually and in different directories, that will be a different story. I am not familiar with `module load mkl`. There are many different scenarios of environment setup, so MKL needs some help to be found.